master/internal/rm/multirm/multirm.go

```diff
@@ -0,0 +1,399 @@
+func (m *MultiRMRouter) ExternalPreemptionPending(allocationID model.AllocationID) error {
+	// MultiRM is currently only implemented for Kubernetes, which doesn't support this.
+	m.syslog.WithError(fmt.Errorf("ExternalPreemptionPending is not implemented for agent, kubernetes, or multi-rm"))
+	return rmerrors.ErrNotSupported
+}
+
+// IsReattachableOnlyAfterStarted routes a IsReattachableOnlyAfterStarted call to a specified resource manager/pool.
+func (m *MultiRMRouter) IsReattachableOnlyAfterStarted(rmName string) bool {
+	resolvedRMName, err := m.getRM(rmName)
+	if err != nil {
+		m.syslog.WithError(err)
+		return false // Not sure what else to return here.
+	}
+
+	return m.rms[resolvedRMName].IsReattachableOnlyAfterStarted(resolvedRMName)
+}
+
+// GetResourcePools returns all resource pools across all resource managers.
+func (m *MultiRMRouter) GetResourcePools() (*apiv1.GetResourcePoolsResponse, error) {
+	res, err := fanOutRMCall(m, func(rm rm.ResourceManager) (*apiv1.GetResourcePoolsResponse, error) {
+		return rm.GetResourcePools()
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	all := &apiv1.GetResourcePoolsResponse{}
+	for _, r := range res {
+		all.ResourcePools = append(all.ResourcePools, r.ResourcePools...)
+	}
+	return all, nil
+}
+
+// GetDefaultComputeResourcePool routes a GetDefaultComputeResourcePool to the specified resource manager.
+func (m *MultiRMRouter) GetDefaultComputeResourcePool(rmName string) (
+	sproto.GetDefaultComputeResourcePoolResponse, error,
+) {
+	resolvedRMName, err := m.getRM(rmName)
+	if err != nil {
+		return sproto.GetDefaultComputeResourcePoolResponse{}, err
+	}
+
+	return m.rms[resolvedRMName].GetDefaultComputeResourcePool(resolvedRMName)
+}
+
+// GetDefaultAuxResourcePool routes a GetDefaultAuxResourcePool to the specified resource manager.
+func (m *MultiRMRouter) GetDefaultAuxResourcePool(rmName string) (sproto.GetDefaultAuxResourcePoolResponse, error) {
+	resolvedRMName, err := m.getRM(rmName)
+	if err != nil {
+		return sproto.GetDefaultAuxResourcePoolResponse{}, err
+	}
+
+	return m.rms[resolvedRMName].GetDefaultAuxResourcePool(resolvedRMName)
+}
+
+// ValidateResourcePool routes a ValidateResourcePool call to the specified resource manager.
+func (m *MultiRMRouter) ValidateResourcePool(rmName string, rpName string) error {
+	resolvedRMName, err := m.getRM(rmName)
+	if err != nil {
+		return err
+	}
+
+	return m.rms[resolvedRMName].ValidateResourcePool(resolvedRMName, rpName)
+}
+
+// ResolveResourcePool routes a ResolveResourcePool request for a specific resource manager/pool.
+func (m *MultiRMRouter) ResolveResourcePool(rmName string, req sproto.ResolveResourcesRequest) (
+	resourceManager, resourcePool string, err error,
+) {
+	resolvedRMName, err := m.getRM(rmName)
+	if err != nil {
+		return rmName, req.ResourcePool, err
+	}
+
+	return m.rms[resolvedRMName].ResolveResourcePool(resolvedRMName, req)
```

Comment:
where does the defaulting pool look up take place?

For example if I have a resource manager with

name=azure with ```azurea, azureb``` pool names

name=gcp with ```gcpa, gcpb```

and I specify ```gcpa``` do we lookup the resource manager by pool name?

Reply:
see the buildRM() func in core.go -- that sets the defaultRMName for the multiRMRouter & has the logic of creating the rm map
If you only specify `gcpa` _but_ you pass in an empty RMName "" _and_ the defaultRMName is `gcp`, then it'll successfully find the pool in the resource manager.
We don't look up resource manager by pool name. If we get an empty rm name we'll default to the defaultRMName, otherwise we error.

Reply:
ugh you're right -- but I don't like the concession we made for that. Then why are we defining a default RM name, if we're going to search through all resource managers anyway? 
(default RM did seem like a shortcut for this tbh)